src/sys/fallocate.rs

```diff
@@ -0,0 +1,20 @@
+use libc;
+use libc::{off_t, c_int};
+use std::os::unix::io::RawFd;
+use {Errno, Result};
+
+libc_bitflags!(
+    pub struct FallocateFlags: libc::c_int {
+        FALLOC_FL_KEEP_SIZE;
+        FALLOC_FL_PUNCH_HOLE;
+        FALLOC_FL_COLLAPSE_RANGE;
+        FALLOC_FL_ZERO_RANGE;
+        FALLOC_FL_INSERT_RANGE;
+        FALLOC_FL_UNSHARE_RANGE;
+    }
+);
+
+pub fn fallocate(fd: RawFd, mode: c_int, offset: off_t, len: off_t) -> Result<c_int> {
```

Comment:
The datatypes for this should not be the raw libc types. For example, `mode` should be `FallocateFlags`. I don't know what we should do for `off_t`, but it's likely we'd want to use `i64`s here instead. The main thing here is that we want to restrict the number of possible invalid inputs by using the type system which gives us compile-time errors. Like reading through the docs you can see that there are invalid combinations of `FALLOC_*` flags. We might want to figure out a way to encode that (I haven't thought about it too hard, but maybe)...

Reply:
Thanks for the feedback. I will implement the requested changes and get back in case of a query.

Reply:
@Susurrus How can I leverage the type system to restrict the number of possible valid inputs? I asked a friend of mine on how to do this, and he said that I would have to leverage traits.

Reply:
@SanchayanMaity no need to use traits.  In this case, all you have to do is change `mode`'s type to `FallocateFlags`, like we do for `SigAction::new`.  I don't have any good ideas about how to restrict the combinations of flags besides declaring an `enum` with every valid combination.  But that would be awkward, because the values of the `enum` would no longer match anything in the `fallocate(2)` man page.

Reply:
@asomers Alright.
@Susurrus  The diff is what I have now. Does this look good now? Had one doubt. I have send patches before, but, for this pull request do I do a force push to the same branch of my fork or create a new branch and then create another pull request?

```C
diff --git a/CHANGELOG.md b/CHANGELOG.md
index d8be190..24c2007 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -30,6 +30,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 - Added socket option variant that enables the timestamp socket
   control message: `nix::sys::socket::sockopt::ReceiveTimestamp`
   ([#663](https://github.com/nix-rust/nix/pull/663))
+- Add nix::sys::fallocate
+  ([#768](https:://github.com/nix-rust/nix/pull/768))
 
 ### Changed
 - Renamed existing `ptrace` wrappers to encourage namespacing ([#692](https://github.com/nix-rust/nix/pull/692))
diff --git a/src/fcntl.rs b/src/fcntl.rs
index 003c316..4ccbf12 100644
--- a/src/fcntl.rs
+++ b/src/fcntl.rs
@@ -341,3 +341,28 @@ pub fn vmsplice(fd: RawFd, iov: &[IoVec<&[u8]>], flags: SpliceFFlags) -> Result<
     Errno::result(ret).map(|r| r as usize)
 }
 
+libc_bitflags!(
+    pub struct FallocateFlags: libc::c_int {
+        /// File size will not be changed even if offset + len is greater than file size.
+        FALLOC_FL_KEEP_SIZE;
+        /// Deallocates space in byte range starting at offset and continuing for len bytes.
+        FALLOC_FL_PUNCH_HOLE;
+        /// Removes a byte range from a file without leaving a hole. The byte range to
+        /// collapse starts at offset and continues for len bytes.
+        FALLOC_FL_COLLAPSE_RANGE;
+        /// Zeroes space in byte range starting at offset and continuing for len bytes.
+        FALLOC_FL_ZERO_RANGE;
+        /// Increases file space by inserting a hole within the file size and does not
+        /// overwrite existing data. Hole starts at offset and continues for len bytes.
+        FALLOC_FL_INSERT_RANGE;
+        /// Shared file data extants will be made private to the file to guarantee that a
+        /// subsequent write will not fail due to lack of space.
+        FALLOC_FL_UNSHARE_RANGE;
+    }
+);
+
+#[cfg(any(target_os = "linux", target_os = "android"))]
+pub fn fallocate(fd: RawFd, mode: FallocateFlags, offset: i64, len: i64) -> Result<c_int> {
+    let res = unsafe { libc::fallocate(fd, mode.bits(), offset, len) };
+    Errno::result(res)
+}
```

Reply:
That's definitely better, and you can do whatever style you want, I don't care, but I will ask to have your commit squashed into a single one at the end.

One thing to note about doc comments is that they should be written with a summary sentence on it's own line, then a blank line, then more details. I really encourage you to read all of the [documentation chapter of the Rust book](https://doc.rust-lang.org/stable/book/first-edition/documentation.html) to get familiar with the general usage of doc comments.

You still need a doccomment here for the `fallocate` function and the `FallocateFlags` struct as a whole (see the `src/macros.rs` file for where it should go.

Reply:
Also, I think this is still supposed to be in the `fcntl.rs` file as that's the same header that defines it in `libc`. If you confirm that's the case, then please move it there.

Reply:
@Susurrus You are referring to rust-libc right? fallocate seems to be defined in [mod.rs](https://github.com/rust-lang/libc/blob/master/src/unix/notbsd/linux/mod.rs#L1115). What do you mean by "You still need a doccomment here for the fallocate function and the FallocateFlags struct as a whole (see the src/macros.rs file for where it should go."?

Reply:
You need a comment describing the `FallocateFlags` type. You have comments for every variant, but not for the overall type. If you run `cargo doc --no-deps --open` and then look at the `fcntl.rs` page you'll see that the `FallocateFlags` type won't have any documentation. That should be fixed.